src/cursor.rs

```diff
@@ -800,15 +797,18 @@ impl GreenNode {
         &self,
         end_index: usize,
         mut offset: TextUnit,
-    ) -> impl Iterator<Item = (&GreenElement, (usize, TextUnit))> {
-        self.children()[..end_index].iter().rev().enumerate().map(move |(index, element)| {
-            offset -= element.text_len();
-            (element, (end_index - index - 1, offset))
-        })
+    ) -> impl Iterator<Item = (GreenElementRef<'_>, (usize, TextUnit))> {
+        // FIXME: uhoh: reverse iteration?
+        //        self.children().take(end_index).rev().enumerate().map(move |(index, element)| {
+        //            offset -= element.text_len();
+        //            (element, (end_index - index - 1, offset))
+        //        })
+        unimplemented!();
+        vec![].into_iter()
```

Comment:
@matklad, You suggested the variable length encoding, so perhaps you can suggest the best way to deal with this.

`children_to` is used in `last_child`, `last_child_or_token`, `prev_sibling`, and `prev_sibling_or_token`. If we do go with this layout, either the cursor has to keep a previous pointer or previous becomes a decently more expensive operation.

Reply:
Hm, would a *single* previous pointer actually help in this case? You'd be able to go one step backwards, but then you wouldn't have previous-previous pointer. 

I think we should store a byte-sized tag between the elements, which describes prev/next elements. *maybe* we should even use XOR-linked list trick for the tag, but I doubt it'll be actually important.

Reply:
The tag is necessarily `u64` big for alignment, so we've a good deal of space to stick extra info. If I'm reading you right, then instead of the tag being 0u64 and 1u64, it should be 0u32 and 1u32, but we also provide a tag for the previous kind to allow for reverse iteration. (Or equivalently any other packing of two bits into 64.)

That's actually pretty clever and would work perfectly!

If we wanted to get crazy each tag could store 32 (if we want an end marker) or even 64 kinds for larger strides, but that definitely seems like overkill.

Reply:
Found an issue while trying to implement it: we need a way to find the end to start with. And the header of the `GreenNode` is full up with `AtomicUsize, TextUnit(u32), SyntaxKind(u16), child_count(u16)`.

We can give up `ExactSizeIterator` and store `fam_length(u16)` instead to allow reverse iteration. I'm going to implement that approach.

Reply:
Ugh, right, we can't have `.rev().enumerate()` without `DoubleEndedIterator + ExactSizeIterator`.

Time for me to examine what's actually being done and try to find a way around wanting both `DoubleEndedIterator` _and_ `ExactSizeIterator`.